web/src/sessions.rs

```diff
@@ -0,0 +1,115 @@
+use axum::{
+    body::Body,
+    extract::FromRequestParts,
+    http::{self},
+    response::{IntoResponse, Response},
+};
+use cja::{app_state::AppState as _, tower_cookies::Cookies};
+
+use uuid::Uuid;
+
+use crate::AppState;
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct Session {
+    #[allow(clippy::struct_field_names)]
+    pub session_id: Uuid,
+    pub user_id: Option<Uuid>,
+    pub expires_at: chrono::DateTime<chrono::Utc>,
+    pub logged_out_at: Option<chrono::DateTime<chrono::Utc>>,
+    pub updated_at: chrono::DateTime<chrono::Utc>,
+    pub created_at: chrono::DateTime<chrono::Utc>,
+}
+
+impl Session {
+    async fn create(state: &AppState, cookies: &Cookies) -> Result<Self, SessionError> {
+        let session_id = Uuid::new_v4();
+        let expires_at = chrono::Utc::now() + chrono::Duration::days(7);
+
+        let session = sqlx::query_as!(
+            Session,
+            r"
+            INSERT INTO Sessions (session_id, expires_at)
+            VALUES ($1, $2)
+            RETURNING *
+            ",
+            session_id,
+            expires_at
+        )
+        .fetch_one(state.db())
+        .await
+        .map_err(|_| SessionError)?;
+
+        let private = cookies.private(state.cookie_key());
+
+        let session_cookie =
+            cja::tower_cookies::Cookie::build(("session_id", session.session_id.to_string()))
+                .path("/")
+                .http_only(true)
+                .secure(true)
+                .expires(None);
+        private.add(session_cookie.into());
+
+        Ok(session)
+    }
+}
+
+pub struct SessionError;
+
+impl IntoResponse for SessionError {
+    fn into_response(self) -> Response<Body> {
+        http::StatusCode::INTERNAL_SERVER_ERROR.into_response()
+    }
+}
+
+#[async_trait::async_trait]
+impl FromRequestParts<AppState> for Session {
```

Comment:
This is how we let you add `session: Session` to your handlers and axum automatically knows what to do!

Well what it does is call this `from_request_parts` method to make a `Session` for you before calling your handler!

This code looks in the Users cookies for a `session_id` and uses that to pull the right session from the DB

Reply:
Is this polymorphism? Also what's the reason for creating a new `instance` of Session instead of making another `async `function inside the `Session Trait`?

Reply:
No this isn't really polymorphism, I think Axum calls these `extractors`. Basically we can 'extract' some data from the requests (and in our case the DB helps too!)

> Also what's the reason for creating a new instance of Session instead of making another async function inside the Session Trait? 

First just some semantics! `Session` here is a struct, and the `FromRequestParts` is the trait we are implementing for our `Session` struct!

So we want an instance of a Session struct so that it can have all its data! And we need an instance of the struct to call methods on it that take a `self` parameter, or just to access the data they hold!

And in our request handlers, we want to know what specific Session this web request is associated with. We care about the _specific_ session_id not just the general concept of a Session, so we want an instance!

Does that kinda make sense?